File: desktop/Dialog.cpp
/* Dialog.cpp
Copyright 2020 Michael Zahniser
*/

#include "Dialog.h"

#include "Color.h"
#include "Sprite.h"
#include "Text.h"
#include "Variables.h"
#include "World.h"

#include <map>

using namespace std;

namespace {
	// This class stores a single "node" of the dialog, i.e. a section of the
	// conversation that you can skip to via a goto or an option.
	class Node : public vector<string> {
	public:
		string ask;
	};
	map<string, Node> nodes;
	
	// Parameters for drawing dialogs:
	const Color DIALOG_COLOR(180, 180, 180);
	const Color OPTION_COLOR(200, 200, 200);
	const Color HOVER_COLOR(150, 170, 200);
	const Color LINE_COLOR(0, 0, 0);
	const int WRAP_WIDTH = 400;
	const Point BOX_PAD(10, 10);
	const int OPTION_PAD = 10 + 2 * BOX_PAD.Y();
	const int DIALOG_Y = 40 + BOX_PAD.Y();
	const Point TEXT_OFFSET(0, 2);
	
	// Skip an indented block in a data file.
	void SkipBlock(Data &data)
	{
		// Skip the indented block after this line. Note that data.Indent()
		// returns 0 if we reach a blank line, so there is no need to check
		// data.Size().
		int indent = data.Indent();
		while(data.Next() && data.Indent() > indent)
			continue;
	}
	// Draw a rectangle with a frame around it.
	void FrameRect(SDL_Surface *surface, Rect rect, const Color &color)
	{
		SDL_FillRect(surface, &rect, LINE_COLOR(surface));
		rect.Grow(-1);
		SDL_FillRect(surface, &rect, color(surface));
	}
}



// Load a dialog definition from the given data file.
void Dialog::Load(Data &data)
{
	Node &node = nodes[data.Value()];
	while(data.Next() && data.Size())
	{
		if(data.Tag() == "ask")
			node.ask = data.Value();
		else
			node.push_back(data.Line());
	}
}



Dialog::Dialog(World &world)
	: world(world)
{
}



// Begin the given dialog.
void Dialog::Begin(const string &name)
{
	data = Data(nodes[name]);
	Step();
}



void Dialog::Begin(const vector<string> &changes)
{
	data = Data(changes);
	Step();
}



// Write the dialog's current state to a saved game file.
void Dialog::Save(ostream &out) const
{
	// Write the current state that the dialog has accumulated.
	if(!text.empty())
		out << "say " << text << '\n';
	if(icon)
		out << "icon " << icon << '\n';
	if(scene)
		out << "scene " << scene << '\n';
	for(const string &option : options)
		out << "option " << option << '\n';
	if(!exitText.empty())
		out << "exit " << exitText << '\n';
	// Also write the remainder of this node that it has not yet executed.
	data.Save(out);
}



void Dialog::Close()
{
	ClearOptions();
	icon = 0;
	scene = 0;
}



// Check if the dialog needs to be drawn.
bool Dialog::IsOpen() const
{
	return !text.empty() || icon || scene || !optionText.empty();
}



// Draw the dialog. If the given mouse position is on one of the options,
// highlight it.
void Dialog::Draw(SDL_Surface *screen, Point hover) const
{
	Text wrap(WRAP_WIDTH);
	wrap.Wrap(text);

	// The dialog box is laid out as follows:
	// First, the scene (if any), centered horizontally and occupying the whole
	// width of the box. (If necessary, the box expands to fit it.)
	// Then, the icon and the text side by side, with BOX_PAD between them. If
	// there is extra width (because of a wide scene) they are centered.
	// TODO: Is there any way to simplify this math?
	Point dialogSize(wrap.Width(), wrap.Height());
	if(icon)
	{
		const Sprite &sprite = Sprite::Get(icon);
		dialogSize.X() += Sprite::Get(icon).Width() + BOX_PAD.X();
		dialogSize.Y() = max(dialogSize.Y(), sprite.Height());
	}
	if(scene)
	{
		const Sprite &sprite = Sprite::Get(scene);
		dialogSize.X() = max(dialogSize.X(), sprite.Width());
		dialogSize.Y() += sprite.Height() + BOX_PAD.Y();
	}
	// Center the dialog in X, and put it near the top of the screen in Y.
	Point corner((screen->w - dialogSize.X()) / 2, DIALOG_Y);
	Rect rect(corner - BOX_PAD, corner + dialogSize + BOX_PAD);
	FrameRect(screen, rect, DIALOG_COLOR);
	
	// Draw the scene, icon, and text.
	if(scene)
	{
		const Sprite &sprite = Sprite::Get(scene);
		// The top left corner of the sprite should be at the top of the box and
		// the left and right should have equal padding.
		int hPad = (dialogSize.X() - sprite.Width()) / 2;
		sprite.Draw(screen, corner - sprite.Bounds().TopLeft() + Point(hPad, 0));
		int height = BOX_PAD.Y() + sprite.Height();
		corner.Y() += height;
		dialogSize.Y() -= height;
	}
	int width = wrap.Width();
	if(icon)
	{
		const Sprite &sprite = Sprite::Get(icon);
		width += sprite.Width() + BOX_PAD.X();
		int hPad = (dialogSize.X() - width) / 2;
		sprite.Draw(screen, corner - sprite.Bounds().TopLeft() + Point(hPad, 0));
	}
	// Given the width of this entire lower section, calculate the text offset.
	corner.X() += (dialogSize.X() + width) / 2 - wrap.Width();
	wrap.Draw(screen, corner + TEXT_OFFSET);
	// Options should only be as wide as the text.
	dialogSize.X() = wrap.Width();
	
	// Show the options, or else a prompt to continue.
	static const vector<string> PROMPT = {"(Click anywhere to continue.)"};
	string number = "0: ";
	optionRects.clear();
	for(const string &option : optionText.empty() ? PROMPT : optionText)
	{
		corner.Y() += dialogSize.Y() + OPTION_PAD;
		++number[0];
		wrap.Wrap(number + option);
		dialogSize.Y() = wrap.Height();
		rect = Rect(corner - BOX_PAD, corner + dialogSize + BOX_PAD);
		FrameRect(screen, rect, rect.Contains(hover) ? HOVER_COLOR : OPTION_COLOR);
		wrap.Draw(screen, corner + TEXT_OFFSET);
		optionRects.push_back(rect);
	}
}



// Handle an event, and return true if the screen needs to be redrawn.
bool Dialog::Handle(const SDL_Event &event)
{
	if(event.type == SDL_MOUSEBUTTONDOWN)
	{
		if(optionText.empty())
		{
			Acknowledge();
			return true;
		}
		size_t clicked = Button(Point(event.button.x, event.button.y));
		if(clicked != optionRects.size())
		{
			Choose(clicked);
			return true;
		}
	}
	else if(event.type == SDL_MOUSEMOTION)
	{
		Point point(event.motion.x, event.motion.y);
		Point previous = point - Point(event.motion.xrel, event.motion.yrel);
		if(Button(point) != Button(previous))
			return true;
	}
	else if(event.type == SDL_KEYDOWN)
	{
		SDL_Keycode sym = event.key.keysym.sym;
		if(!optionText.empty() && sym >= '1' && sym <= '9')
		{
			size_t choice = sym - '1';
			if(choice < optionText.size())
			{
				Choose(choice);
				return true;
			}
		}
		else if(optionText.empty() && (sym == '1' || sym == SDLK_SPACE || sym == SDLK_RETURN))
		{
			Acknowledge();
			return true;
		}
	}
	return false;
}



// Select the given option, advancing the dialog to the next page.
void Dialog::Choose(int option)
{
	// Check if the chosen option was to exit the conversation.
	if(static_cast<size_t>(option) < options.size())
	{
		visited.insert(options[option]);
		data = Data(nodes[options[option]]);
		ClearOptions();
		Step();
	}
	else
		Close();
}



// If the dialog is just waiting for acknowledgment, instead of for a
// specific choice, go on to the next page.
void Dialog::Acknowledge()
{
	if(!data)
		Close();
	else
		Step();
}



void Dialog::Step()
{
	// The scene resets every page. The icon persists.
	scene = 0;
	bool spoke = false;
	while(data)
	{
		// Special case: a "goto" node moves us to new data immediately.
		if(data.Tag() == "goto")
		{
			data = Data(nodes[data.Value()]);
			// Process this line instead of advancing to the next one.
			continue;
		}
		else if(data.Tag() == "if")
		{
			if(!Variables::Eval(data.Value()))
			{
				SkipBlock(data);
				// We are now at the first line after the end of the "if" block
				// that we skipped. If that line is an "else" we can skip that
				// line and move on into the contents of the else. Otherwise, we
				// need to "continue" to avoid advancing past this line.
				if(data.Tag() != "else")
					continue;
			}
		}
		else if(data.Tag() == "else")
		{
			// If we got to this line, it means we didn't skip the previous "if"
			// clause, so we need to skip this one.
			SkipBlock(data);
			// We don't need to call data.Next() because it is already at the
			// first line after the block.
			continue;
		}
		else if(data.Tag() == "option")
		{
			if(!visited.count(data.Value()))
				options.push_back(data.Value());
		}
		else if(data.Tag() == "exit")
			exitText = data.Size() > 1 ? data.Value() : "(End conversation.)";
		else if(data.Tag() == "icon")
			icon = data[1];
		else if(data.Tag() == "scene")
			scene = data[1];
		else if(data.Tag() == "add")
		{
			string room = data.Value();
			int indent = data.Indent();
			data.Next();
			while(data.Indent() > indent)
			{
				if(data.Tag() == "interaction")
					world.Add(Interaction(data), room);
				else
				{
					world.Add(data[0], data[1], data.Value(2), room);
					data.Next();
				}
			}
			// Return to the top of the loop without calling data.Next() again.
			continue;
		}
		else if(data.Tag() == "remove")
		{
			string room = data.Value();
			int indent = data.Indent();
			while(data.Next() && data.Indent() > indent)
				world.Remove(data.Value(0), room);
			// Return to the top of the loop without calling data.Next() again.
			continue;
		}
		else if(data.Tag() == "enter")
			world.Enter(data[1], data[2]);
		else if(data.Tag() == "face")
			world.Face(data[1]);
		else if(data.Tag() == "set")
			Variables::Set(data.Value());
		else if(data.Tag() == "say")
		{
			// If we've reached another "say" line, that means the current
			// "paragraph" should be shown immediately, and we don't need to
			// show options because we're not yet at the end of the stream.
			if(spoke)
				return;
			spoke = true;
			text = data.Value();
		}
		// If we didn't "continue" up above, advance to the next line.
		data.Next();
	}
	
	// We've reached the end of the stream. Show any options we accumulated.
	for(const string &option : options)
		optionText.push_back(nodes[option].ask);
	if(!exitText.empty())
		optionText.push_back(exitText);
}



void Dialog::ClearOptions()
{
	// Begin accumulating options from scratch.
	optionText.clear();
	options.clear();
	exitText.clear();
	// The text disappears, but the icon will remain until cleared manually or
	// until the conversation ends.
	text.clear();
}



// Check if the given point is inside one of the options. If not, this
// returns the size of the options vector.
size_t Dialog::Button(Point p)
{
	for(size_t i = 0; i < optionRects.size(); ++i)
		if(optionRects[i].Contains(p))
			return i;
	return optionRects.size();
}

File: web/dialog.js
import {Data} from "./data.js";
import {Interaction} from "./interaction.js";
import {Sprite} from "./sprite.js";
import {Style} from "./style.js";
import {Variables} from "./variables.js";
import {World} from "./world.js";

/** An object for parsing and displaying dialog in the game.
 * @typedef {Object} Dialog
 */
export class Dialog {
  /** Load a dialog definition from the given data file.
   * @param {Data} data
   */
  static load(data) {
    const node = Dialog._nodes[data.value()] = new Node();
    while (data.next() && data.size()) {
      if (data.tag() == "ask") node.ask = data.value();
      else node.lines.push(data.line());
    }
  }

  /** Construct a dialog object that can make changes to the given World.
   * @param {World} world
   */
  constructor(world) {
    this._world = world;
  }

  /** Start the dialog with the given name, if it exists. If the dialog does not produce
   * any visible output, it will run immediately and return.
   * @param {string} name
   */
  begin(name) {
    if (name in Dialog._nodes) {
      this._data = new Data(Dialog._nodes[name].lines);
      this._step();
      if (this.isOpen()) {
        this._keyListener = event => this._keypress(event);
        document.addEventListener("keypress", this._keyListener);
      }
    }
  }

  /** Close the dialog, if it is open. */
  close() {
    this._clearOptions();
    this._optionText.length = 0;
    document.getElementById("options").innerHTML = "";
    this._icon = 0;
    this._scene = 0;
    this._display();
    if (this._keyListener) {
      document.removeEventListener("keypress", this._keyListener);
      this._keyListener = null;
    }
  }

  /** If the dialog is in the "click to continue" state, choose the only option. */
  continue() {
    if (this.isOpen() && this._optionText.length == 1) this._choose(0);
  }

  /** Check if the dialog is open, i.e. something is being displayed.
   * @returns {boolean} True if the current dialog state requires displaying anything.
   */
  isOpen() {
    return this._text.length || this._icon || this._scene || this._optionText.length;
  }

  // Step the dialog forward to the next "say" block.
  _step() {
    // Remove any options currently being displayed.
    this._optionText.length = 0;
    document.getElementById("options").innerHTML = "";
    // The "scene" resets with each page of the dialog. The icon does not.
    this._scene = 0;
    let spoke = false;
    while (this._data.isValid()) {
      if (this._data.tag() === "goto") {
        const name = this._data.value();
        if (name in Dialog._nodes) {
          this._data = new Data(Dialog._nodes[name].lines);
        }
        continue;
      }
      else if (this._data.tag() === "if") {
        if (!Variables.eval(this._data.value())) {
          skipBlock(this._data);
          // We are now at the first line after the end of the "if" block
          // that we skipped. If that line is an "else" we can skip that
          // line and move on into the contents of the else. Otherwise, we
          // need to "continue" to avoid advancing past this line.
          if (this._data.tag() !== "else") continue;
        }
      }
      else if (this._data.tag() === "else") {
        // If we got to this line, it means we didn't skip the previous "if"
        // clause, so we need to skip this one.
        skipBlock(this._data);
        // We don't need to call data.next() because it is already at the
        // first line after the block.
        continue;
      }
      else if (this._data.tag() === "end") {
        if (spoke) break;
        this._world.end();
        return;
      }
      else if (this._data.tag() === "option") {
        const name = this._data.value();
        if (name in Dialog._nodes && !this._visited.has(Dialog._nodes[name])) {
          this._options.push(Dialog._nodes[name]);
        }
      }
      else if (this._data.tag() === "exit") {
        this._exitText = (this._data.size() > 1 ? this._data.value() : "(End conversation.)");
      }
      else if (this._data.tag() === "icon") {
        this._icon = +this._data.arg(1);
      }
      else if (this._data.tag() === "scene") {
        this._scene = +this._data.arg(1);
      }
      else if (this._data.tag() == "add") {
        const room = this._data.value();
        const indent = this._data.indent();
        this._data.next();
        while (this._data.indent() > indent) {
          if (this._data.tag() == "interaction") {
            this._world.addInteraction(new Interaction(data), room);
          }
          else {
            this._world.addSprite(+data.arg(0), data.point(1), data.value(2), room);
            this._data.next();
          }
        }
        // Return to the top of the loop without calling data.next() again.
        continue;
      }
      else if (this._data.tag() == "remove") {
        const room = this._data.value();
        const indent = this._data.indent();
        while (this._data.next() && this._data.indent() > indent) {
          this._world.remove(this._data.value(0), room);
        }
        // Return to the top of the loop without calling data.next() again.
        continue;
      }
      else if (this._data.tag() == "enter") {
        this._world.enter(this._data.point(1), this._data.arg(2));
      }
      else if (this._data.tag() == "face") {
        this._world.face(+this._data.arg(1));
      }
      else if (this._data.tag() == "set") {
        Variables.set(this._data.value());
      }
      else if (this._data.tag() == "say") {
        // If we've reached another "say" line, that means the current
        // "paragraph" should be shown immediately, and we don't need to
        // show options because we're not yet at the end of the stream.
        if (spoke) break;
        spoke = true;

        this._text = this._data.value();
      }
      // If we didn't "continue" up above, advance to the next line.
      this._data.next();
    }

    // We've reached the point where we should update the dialog page.
    // Check if we reached the end of the node, meaning that any options
    // we have accumulated should be displayed.
    if (this._data.isValid()) this._optionText.push("(Click to continue.)");
    else {
      // When accumulating options, we already made sure that they are all valid.
      for (const node of this._options) {
        this._optionText.push(node.ask);
      }
      if (this._exitText) {
        this._optionText.push(this._exitText);
      }
      else if (this._text) {
        this._optionText.push("(Click to continue.)");
      }
    }

    this._display();
  }

  _display() {
    const dialog = document.getElementById("dialog");
    if (!this.isOpen()) {
      dialog.style.display = "none";
      return;
    }
    dialog.style.display = "block";

    // Get the four elements of the document that we can customize.
    const text = document.getElementById("text");
    const icon = document.getElementById("icon");
    const scene = document.getElementById("scene");
    const options = document.getElementById("options");
    // TODO: Avoid hard-coding this width and the padding.
    let width = 400;

    // Set the icon.
    if (this._icon) width += 10 + this._setSprite(icon, this._icon);
    else icon.style.display = "none";
    // Set the scene.
    if (this._scene) width = Math.max(width, this._setSprite(scene, this._scene));
    else scene.style.display = "none";
    // Set the text.
    text.innerHTML = Style.apply(this._text);
    // Add the padding to the width.
    width += 20;
    dialog.style.width = width + "px";
    dialog.style.marginLeft = -0.5 * width + "px";

    // Add the options.
    let index = 0;
    for (const option of this._optionText) {
      // Make a copy of the index for the arrow function to hold onto.
      const i = index++;
      let element = document.createElement("div");
      element.onmousedown = () => this._choose(i);
      element.className = "option box";
      element.innerHTML = Style.apply(index + ": " + option);
      options.append(element);
    }
  }

  _setSprite(element, index) {
    const sprite = Sprite.get(index);
    if (!sprite) return 0;

    element.style.display = "block";
    element.style.width = sprite.width() + "px";
    element.style.height = sprite.height() + "px";
    element.style.background = sprite.source();
    return sprite.width();
  }

  _choose(index) {
    if (index < this._options.length) {
      // Note: before adding theis option to the list, we already checked
      // that it exists and is not visited.
      const dialog = this._options[index];
      this._visited.add(dialog);
      this._data = new Data(dialog.lines);
      this._clearOptions();
      this._step();
    }
    else if (this._data.isValid()) this._step();
    else this.close();
  }

  /** Handle a keyboard event. This listener is only active when the dialog is visible.
   * @param {KeyboardEvent} event
   */
  _keypress(event) {
    const index = +String.fromCharCode(event.charCode);
    if (index >= 1 && index <= 9) this._choose(index - 1);
  }

  _clearOptions() {
    // Begin accumulating options from scratch.
    this._options.length = 0;
    this._exitText = "";
    // The text disappears, but the icon will remain until cleared manually or
	  // until the conversation ends.
	  this._text = "";
  }

  // Dictionary of all the dialog nodes.
  /** @type {Object.<string, Dialog>} */
  static _nodes = {};

  // Hold a reference to the World object, to apply changes to it.
  /** @type {World} */
  _world;
  // The data we are currently iterating through.
  /** @type {Data} */
  _data;
  // Track whether we're listening to key events.
  /** @type {function(KeyboardEvent) => void} */
  _keyListener;

  // Display state:
  /** @type {string} */
  _text = "";
  /** @type {number} */
  _icon = 0;
  /** @type {number} */
  _scene = 0;
  /** @type {string[]} */
  _optionText = [];

  // Remember which dialog nodes we have visited. They will not be presented as
  // an option again until the player leaves the room.
  /** @type {Set<Dialog>} */
  _visited = new Set();
  // These are the nodes that may be "visited" from the current one.
  /** @type {Dialog[]} */
  _options = [];
}


// This class represents a single "dialog" object in the data files.
class Node {
  ask;
  lines = [];
}

// Skip an indented block in a data file.
function skipBlock(data) {
  const indent = data.indent();
  while (data.next() && data.indent() > indent) continue;
}
